modules/http4s/src/smithy4s/http4s/internals/SmithyHttp4sClientEndpoint.scala

```diff
@@ -79,12 +82,16 @@ private[http4s] class SmithyHttp4sClientEndpointImpl[F[_], Op[_, _, _, _, _], I,
   method: org.http4s.Method,
   endpoint: Endpoint[Op, I, E, O, SI, SO],
   httpEndpoint: HttpEndpoint[I],
-  compilerContext: CompilerContext[F]
+  compilerContext: CompilerContext[F],
+  middleware: HttpApp[F] => HttpApp[F]
 )(implicit effect: EffectCompat[F]) extends SmithyHttp4sClientEndpoint[F, Op, I, E, O, SI, SO] {
 // format: on
 
+  private val transformedClient: Client[F] =
+    Client.fromHttpApp[F](middleware(client.toHttpApp))
```

Comment:
Right, so unfortunately, I don't think we should rely on `fromHttpApp` in non-test code, because it's a [non-trivial transformation](https://github.com/http4s/http4s/blob/series/0.23/client/shared/src/main/scala/org/http4s/client/Client.scala#L278) and I'm somewhat scared that we'd be giving a footgun to the users, as it's easy to mess up the request lifecycle in client-side middleware. 

Essentially the main problem is that the HttpApp abstraction should work against `Resource[F, *]` instead of `F`. I believe there's an issue somewhere in http4s, but the impact will be quite heavy on the ecosystem so I think they're waiting for further stabilisation before pulling the trigger on that. 

So, we need to require the middleware to be provided as a `Client[F] => Client[F]`, which unfortunately implies creating another construct to echo the `EndpointSpecificMiddleware`, which may mean that we'd have to rename that middleware to `EndpointSpecificServerMiddleware which is a bit of a mouthful. 

How about `ServerEndpointMiddleware` and `ClientEndpointMiddleware` ?

Reply:
Added in https://github.com/disneystreaming/smithy4s/pull/614/commits/75313f5a407bb6021c304029c64b9a6439e14b0f